### PR TITLE
Checksum

### PR DESCRIPTION
**Pull request by qoire**
## Notice

It is not allowed to submit your PR to the master branch directly, please submit your PR to the master-pre-merge branch.

## Description

Integrates initial checksum code, this will serve as the reference for other implementations (#543). Currently not _used_ in the codebase itself.

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [ ] Bug fix.
- [x] New feature.
- [ ] Enhancement.
- [ ] Unit test.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [x] Requires documentation update.

## Testing

Tested to match the our web3.js 1.0 implementation.

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests for my fix or feature.
- [ ] I have made appropriate changes to the corresponding documentation.
- [ ] My code generates no new warnings.
- [ ] Any dependent changes have been made.

### CODE REVIEW

**Comment by qoire** [comment]
Cool, will do

**Comment by aion-kelvin** [comment]
+1 to what @AionJayT  said

File: modCrypto/test/org/aion/crypto/ChecksumTest.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.aion.crypto;
+
+import org.aion.base.type.Address;
+import org.aion.base.util.ByteUtil;
```
Comment (aionick) [inline modCrypto/test/org/aion/crypto/ChecksumTest.java]: these two imports can be removed

File: modCrypto/src/org/aion/crypto/AddressSpecs.java [hunk 2]
```diff
@@ -25,4 +26,26 @@ private AddressSpecs() {}
         buf.put(HashUtil.h256(publicKey), 1, 31);
         return buf.array();
     }
+
+    public static String checksummedAddress(String address) {
+        assert address != null;
+        address = address.replace("0x", "");
```
Comment (aionick) [inline modCrypto/src/org/aion/crypto/AddressSpecs.java]: If we are given a string with "0x" sitting in the middle of it, or sitting in multiple places, this replacement could strip the string back down to a legitimate address - do we want this kind of behaviour?

Comment (arajasek) [reply modCrypto/src/org/aion/crypto/AddressSpecs.java]: I think this is an interesting question, though I don't have a clear idea of the gravity of the implications.

I assume it's easy enough to "fix", simply only replace starting "0x"s?

File: modCrypto/src/org/aion/crypto/AddressSpecs.java [hunk 3]
```diff
@@ -25,4 +26,31 @@ private AddressSpecs() {}
         buf.put(HashUtil.h256(publicKey), 1, 31);
         return buf.array();
     }
+
+    public static String checksummedAddress(String address) {
```
Comment (aion-kelvin) [inline modCrypto/src/org/aion/crypto/AddressSpecs.java]: Can we change the method signature to return Optional<String>?  And in places where we currently return null, return `Optional.empty()` instead?

File: modCrypto/src/org/aion/crypto/AddressSpecs.java [hunk 4]
```diff
@@ -25,4 +26,31 @@ private AddressSpecs() {}
...
+        assert address.length() == 64;
+        byte[] h;
+        try {
+            h = HashUtil.h256(ByteUtil.hexStringToBytes(address));
+        } catch (Exception e) {
+            e.printStackTrace();
+            return null;
+        }
+        if (h == null) return null; //address is invalid
+        System.out.println(ByteUtil.toHexString(h));
```
Comment (aion-kelvin) [inline modCrypto/src/org/aion/crypto/AddressSpecs.java]: println not needed

File: modCrypto/src/org/aion/crypto/AddressSpecs.java (hunk 2 above)
Comment (aion-kelvin) [reply modCrypto/src/org/aion/crypto/AddressSpecs.java]: I think the "0x not at the start of string" problem still needs to be fixed (also can yuo add a unit test to make sure this doesn't happen?)

Suggestion: `address.replaceFirst("^0x", "")`

https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#replaceFirst(java.lang.String,%20java.lang.String)

File: modCrypto/src/org/aion/crypto/AddressSpecs.java [hunk 5]
```diff
@@ -25,4 +27,31 @@ private AddressSpecs() {}
...
+        for (int i = 0; i < 64; i++) {
+            if (Character.isDigit(caddr[i]))
+                continue;
+
+            if (Character.isAlphabetic(caddr[i])) {
+                caddr[i] = b.get(i) ? Character.toUpperCase(caddr[i]) : Character.toLowerCase(caddr[i]);
+                continue;
+            }
+        }
+        return Optional.of(String.valueOf(caddr));
```
Comment (AlexandraRoatis) [inline modCrypto/src/org/aion/crypto/AddressSpecs.java]: I think it would make sense to append the `0x` to the resulting address.

**Pull request merged**